plugins/resteasy-http/src/main/java/org/radargun/service/RESTEasyHTTPOperations.java

```diff
@@ -0,0 +1,263 @@
+      }
+
+      private class WrappedValue<W> {
+         EntityTag eTag;
+         W value;
+
+         public WrappedValue(EntityTag eTag, W value) {
+            super();
+            this.eTag = eTag;
+            this.value = value;
+         }
+      }
+
+      @Override
+      public boolean containsKey(K key) {
+         if (service.isRunning()) {
+            String target = buildUrl(service.cacheName, key);
+            try {
+               Invocation head = service.getHttpClient().target(target).request().build(HttpMethod.HEAD);
+               head.invoke();
+               return true;
+            } catch (NotFoundException e) {
+               return false;
+            } catch (Exception e) {
+               throw new RuntimeException("RESTEasyHTTPOperations::containsKey request threw exception: " + target, e);
+            }
+         }
+         return false;
+      }
+
+      @Override
+      public void put(K key, V value) {
+         putInternal(key, value, null);
+      }
+
+      private void putInternal(K key, V value, EntityTag eTag) {
+         if (service.isRunning()) {
+            try {
+               String target = buildUrl(service.cacheName, key);
+               Invocation put = null;
+               if (eTag != null) {
+                  put = service.getHttpClient().target(target).request().accept(service.getContentType())
+                        .header("If-Match", eTag.getValue())
+                        .buildPut(Entity.entity(encodeObject(value), service.getContentType()));
+               } else {
+                  put = service.getHttpClient().target(target).request().accept(service.getContentType())
+                        .buildPut(Entity.entity(encodeObject(value), service.getContentType()));
+               }
+               Response response = put.invoke();
+               int status = response.getStatus();
+               String reason = response.getStatusInfo().getReasonPhrase();
+               response.close();
+               if (status != Status.OK.getStatusCode() && status != Status.CREATED.getStatusCode()
+                     && status != Status.NO_CONTENT.getStatusCode()) {
+                  throw new RuntimeException("RESTEasyHTTPOperations.put::Unexpected HttpStatus: " + status + " "
+                        + reason + " for request " + target);
+               }
+            } catch (IOException e) {
+               throw new RuntimeException(e);
+            }
+         }
+      }
+
+      @Override
+      public V getAndPut(K key, V value) {
+         V prevValue = null;
+         EntityTag eTag = null;
+         if (service.isRunning()) {
+            if (containsKey(key)) {
+               HTTPCache<K, V>.WrappedValue<V> wrap = getInternal(key);
+               eTag = wrap.eTag;
+               prevValue = wrap.value;
+            }
+            if (eTag == null) {
+               put(key, value);
+            } else {
+               putInternal(key, value, eTag);
```

Comment:
It is possible to detect that the failure is due to the If-Match header, because the put or remove will return a 412 (Precondition Failed) response code. The question is whether retrying the operation is correct? None of the other operations are retried, so in my opinion, the current code is the correct behavior. What do you think?